invenio_records_resources/services/records/queryparser/query.py

```diff
@@ -61,19 +61,55 @@ class SearchOptions:
             )
     """
 
-    def __init__(self, identity=None, extra_params=None, tree_transformer_factory=None):
+    def __init__(self, identity=None, extra_params=None, tree_transformer_cls=None):
         """Initialise the parser."""
         self.identity = identity
+        self.tree_transformer_cls = tree_transformer_cls
         self.extra_params = extra_params or {}
-        self.tree_transformer_factory = tree_transformer_factory
+        # the pop or {} is needed due to extra_params being passed from the factory
+        # it is possible that e.g. allow_list=None and then it will fail to set()
+        self.mapping = self.extra_params.pop("mapping", None) or {}
+        self.allow_list = self._calculate_allow_list(
+            self.extra_params.pop("allow_list", None), self.mapping
+        )
+        # process fields list
+        self.fields = self._calculate_fields(
+            # get instead of pop since fields could desired for
+            # both querystring and multi_match
+            self.extra_params.get("fields", None),
+            self.allow_list,
+        )
+
+    def _calculate_allow_list(self, allow_list, mapping):
+        """Calculate the allow list."""
+        if allow_list:  # only add the mapping if there is an allow list
+            # mapping should be at least a subset of the allow list
+            return set(allow_list).union(mapping.values())
+        return None
+
+    def _calculate_fields(self, fields, allow_list):
+        """Calculate the list of fields to query."""
+        if not allow_list:
+            return fields
+
+        fields = fields or []
+        to_remove = set()
+        for field in fields:  # check duplicated fields
+            # FIXME: it does not support fields with wildcards
+            field_name = field.split("^")[0]  # remove potential boosting
+            if field_name in self.allow_list:
```

Comment:
I think that here or we use the param (better) or we user `self.allow_list`:
```suggestion
            if field_name in allow_list:
```